### DISCUSSION TITLE
Where does Kmonad intervenes in the keyboard signal -> GUI stack ?

### DISCUSSION THREAD

**Discussion by t-cadet** [kmonad/kmonad#477] [Q&A] [answered]
At first I thought Kmonad would be mapping a physical position on the keyboard (a scan code ?) to a keycode but it seems to be mapping a keycode to another keycode. In particular this means that if we have 2 computers with the same physical keyboard but with a different layout set in the OS options (tested on windows, say qwerty and azerty), using the same Kmonad config file on both computers will result in a different physical key to keycode mapping.

From that observation I think what happens is roughly:

keyboard -> (scan code) -> OS -> (key code, scan code) -> Kmonad -> (key code) -> app

I’m pretty new to keyboard layouts (I’m not even sure I have the right definition for key and scan codes) and I’m trying to understand what can and cannot be done with Kmonad (or other similar tools). Can someone who’s more advanced explain what is happening in this stack and why Kmonad chose to map key codes rather than scan codes ?

Also, I might be missing some key concepts to setup my layout, so as a side question, what do you think needs to be understood in order to properly configure a keyboard layout ?

**Comment by thoelze1** [comment, marked as answer]
> I thought Kmonad would be mapping a physical position on the keyboard (a scan code ?)

Given two keyboards that look identical, a key on one might send a different scancode than the key in the same position on the other. This would happen, for example, if one keyboard had been programmed at the firmware level (see QMK) as QWERTY and the other as AZERTY. In general, there's no way even for the OS to determine the physical location of a key because keyboards send scancodes rather than some kind of "coordinates" that indicate the physical location of the keypress.

Having keyboards describe to the operating system the location of a keypress (e.g. "topmost row, rightmost key") would be pretty complicated. Providing instead the logical meaning of the pressed key (e.g. backspace) is simpler, and in a way provides the same information: if you know the logical layout of the keyboard (e.g. ANSI, or something you customized yourself in QMK), and you know which logical key was pressed (e.g. backspace), then you can determine which physical key was pressed (e.g. the rightmost key in the topmost row). So these two ways of reporting keypresses amount to the same information, but the latter is simpler and is the one that's used.

> if we have 2 computers with the same physical keyboard but with a different layout set in the OS options (tested on windows, say qwerty and azerty), using the same Kmonad config file on both computers will result in a different physical key to keycode mapping

In theory, this behavior is simply OS-dependent. It would make sense that any modifications you provide to the OS should be performed before any other modifications (e.g. by kmonad) but I can imagine that another operating system might not choose this order. Other projects like Karabiner advise users not to use any OS-level modifications (like QWERTY to AZERTY) while remapping (I'm not sure if kmonad offers advice on combining OS-level remapping). My perspective is that for simplicity's sake you should do all of your remapping with one tool - yes, you could do part of it somewhere else (e.g. at the OS level) but then you have to consider how these two remappings compose. The more common case of remapping occurring at two different levels is with a key repeat interval, where a held non-modifier key is remapped by the OS into a sequence of repeated presses of said key until release. On Mac and Linux (and surely on Windows?), kmonad intercepts keypresses before the OS applies key repeat - the alternative would be very ugly.

Ultimately, because the same keyboard should work across multiple operating systems, there should be a bijection between scancode (the piece of data sent by the keyboard to the OS to indicate a keypress) and its interpretation by the OS (this is what you mean by keycode?). So, these two things are the same.

> From that observation I think what happens is roughly:
> 
> keyboard -> (scan code) -> OS -> (key code, scan code) -> Kmonad -> (key code) -> app

Unless some remapping is being done at the OS level, a scan code is always going to correspond to the same interpretation, so there would be no need for the operating system to provide both to a user application (such as kmonad) as shown above. What actually happens (on Mac) is that the OS translates a scancode received from a keyboard into an internal representation that it uses for key codes, then makes this number available to kmonad, which can intercept that number from propagating to other applications, and then send a (possibly different number, or multiple numbers, or no number) to the OS which then propagates to other programs:

keyboard -> scancode -> OS -> (OS representation) -> kmonad -> (OS representation) -> OS -> app

> what do you think needs to be understood in order to properly configure a keyboard layout ?

All that needs to be understood is the current "map" of your keyboard (more than likely, this just means understanding that each key is labeled with what it does) and the "map" that you want, e.g. AZERTY. Then have a look in the `/keymap` folder at some of the template configurations (these are "trivial" or "clear" remaps that will cause your keyboard to function exactly as it does naturally) or some of the user configurations (these ones will  ~change~ improve how the keyboard behaves)

**Comment by t-cadet** [comment]
Thanks for the explanations !

>  My perspective is that for simplicity's sake you should do all of your remapping with one tool

I completely agree with that, it actually didn’t occur to me that the layout could be disabled on windows, I’ll have to search how to do that. Actually I thought of the layout as being part of the OS, and so on my diagram the "key code" was the scan code after it had been translated by the layout.

It would be nice if Kmonad had an option to bypass other remapping programs: ideally I want to put my layout on a USB drive and be able to use it on any computer, possibly without admin rights.
